src/core/DevAI.ts

```diff
@@ -0,0 +1,240 @@
+     * @param tag - The tag whose content you want to extract.
+     * @returns The content inside the specified tag, or null if the tag isn't found.
+     */
+    private extractTagContent(content: string, tag: string): string | null {
+        const regex = new RegExp(`<${tag}>(.*?)<\/${tag}>`);
+        const match = content.match(regex);
+        return match ? match[1] : null;
+    }
+
+    /**
+     * Executes a CLI command in the terminal and logs the result.
+     * @param command - The command to execute.
+     * @returns A boolean indicating success or failure.
+     */
+    private async executeCommandTool(command: string): Promise<boolean> {
+        try {
+            const terminal = vscode.window.createTerminal("DevAI Terminal"); // Create a new terminal in VS Code.
+            terminal.show(); // Show the terminal.
+            terminal.sendText(command); // Send the command to the terminal.
+            await delay(2000); // Simulate command execution time.
+
+            // Add the success result to the conversation history.
+            this.apiConversationHistory.push({
+                role: "assistant",
+                content: formatToolResult(`Command executed: ${command}`)
+            });
+
+            return true;
+        } catch (error) {
+            // Log the error if the command fails.
+            this.apiConversationHistory.push({
+                role: "assistant",
+                content: formatToolError(`Failed to execute command: ${command}`)
+            });
+            return false;
+        }
+    }
+
+    /**
+     * Reads the content of a file from the given path and logs the result.
+     * @param filePath - The path of the file to read.
+     * @returns A boolean indicating success or failure.
+     */
+    private async readFileTool(filePath: string): Promise<boolean> {
+        try {
+            const absolutePath = path.resolve(cwd, filePath); // Resolve the absolute path of the file.
+            const content = await fs.readFile(absolutePath, "utf8"); // Read the file's content.
+
+            // Log the file content to the conversation history.
+            this.apiConversationHistory.push({
+                role: "assistant",
+                content: formatReadFileContent(filePath, content)
+            });
+
+            return true;
+        } catch (error) {
+            // Log the error if reading the file fails.
+            this.apiConversationHistory.push({
+                role: "assistant",
+                content: formatToolError(`Failed to read file: ${filePath}`)
+            });
+            return false;
+        }
+    }
+
+    /**
+     * Writes content to a file at the given path and logs the result.
+     * @param filePath - The path of the file to write to.
+     * @param content - The content to write to the file.
+     * @returns A boolean indicating success or failure.
+     */
+    private async writeFileTool(filePath: string, content: string): Promise<boolean> {
```

Comment:
We would need to add the functionality to create a new file if it doesn't exist before writing to the file.